api/crunch.ts

```diff
@@ -0,0 +1,224 @@
+
+                genreItem.tracks.push(trackItem);
+                groupItem.all_tracks.push(trackItem);
+            }
+
+            sortByScore(genreItem.tracks);
+            genreItem.score = getGenreScore(groupItem, genreItem.tracks)
+        }
+
+        sortByScore(groupItem.all_tracks);
+    }
+
+    return result;
+}
+
+function getGenreScore(group: GroupListItem, tracks: ScoredTrack[]) {
+    var sum = tracks.reduce((sum, track) => sum + track.score, 0);
+    return sum / group.members.length;
+}
+
+function computePlaylists(data: GroupListItem[]) {
+    const statsNew: Playlist[] = [];
+    for (let group of data) {
+        statsNew.push({ groupId: group.id, name: "Top 10", score: 1, tracks: group.all_tracks.slice(0, 10) });
+
+        for (let genre of group.genres) {
+            statsNew.push({ groupId: group.id, name: genre.name, score: genre.score, tracks: genre.tracks.slice(0, 10) });
+        }
+    }
+    return statsNew;
+}
+
+function getTrackScore(group: Group, track: Track) {
+    return track.plays.map((trackPlays) => {
+        const totalPlays = group.members[trackPlays.user];
+        return trackPlays.plays / totalPlays;
+    }).reduce((sum, score) => sum + score, 0);
+}
+
+function sortByScore(tracks: ScoredTrack[]) {
+    return tracks.sort((a, b) => b.score - a.score);
+}
+
+async function createPlaylist(playlist: Playlist) {
+    await prisma.playlists.create({
+        data: {
+            name: playlist.name,
+            is_current_top_list: true,
+            score: playlist.score,
+            group_id: playlist.groupId,
+            playlist_items: {
+                createMany: {
+                    data: playlist.tracks.map((track) => {
+                        return {
+                            track_id: track.id,
+                            number_of_plays: track.plays.reduce((sum, userPlays) => sum + userPlays.plays, 0),
+                            number_of_unique_listeners: track.plays.length,
+                        }
+                    })
+                }
+            }
+        }
+    });
+}
+
+function fetchPlayedTracksLast7days(): Promise<PlayedTrack[]> {
+    return prisma.$queryRaw<PlayedTrack[]>`
+    SELECT 
+          g.id              AS group_id
+        , u.id              AS user_id
+        , t.id              AS track_id
+        , t.genre           AS genre 
+        , count(p.track_id) AS plays
+    FROM       groups        g 
+    INNER JOIN group_members m ON m.group_id = g.id
+    INNER JOIN users         u ON u.id = m.user_id
+    LEFT  JOIN played_tracks p ON p.user_id = u.id
+    LEFT  JOIN tracks        t ON t.id = p.track_id
+    WHERE p.track_id is null 
+        or (m.entered_at <= p.played_at 
+            and p.played_at >= (current_timestamp() - interval '7 days')
+            and t.genre is not null)
+    GROUP BY 
+          g.id
+        , u.id
+        , t.id
+        , t.genre
+ `;
+}
+
+(BigInt.prototype as any).toJSON = function () {
```

Comment:
Where's this used?

Reply:
I think all of the number from the database come as BigInt, and they have no default json serialization.